### PR TITLE
Add iter_keys_upto_delimiter

### LINKED ISSUE #83: Add 'delimiter' keyword to 'iter_keys'
I would like to use the feature that one can give the keyword `delimiter` to the `store.iter_keys` method (https://github.com/mbr/simplekv/blob/master/simplekv/__init__.py#L102).

Some implementations support this feature:
- azure: https://github.com/Azure/azure-storage-python/blob/master/azure-storage-blob/azure/storage/blob/baseblobservice.py#L1202
- boto: https://github.com/boto/boto/blob/develop/boto/s3/bucket.py#L234

Others would allow an implementation for it:
- redis: Allows pattern (https://redis.io/commands/keys, https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L1260)

For remaining stores one could implement a slow evaluation which iterates over all existing keys.

Are there opinions on this? I would welcome a short discussion about this feature. I am also willing to implement some stuff (don't know how good I can implement the feature for the `db` stores).

### PR DESCRIPTION
**Pull request by crepererum**
PR is split in the following way:

1. fixing travis again
2. implement the generic function
3. implement fast path for FS and Azure

Closes #83

### CODE REVIEW

**Comment by mbr** [comment]
I'm not entirely sure this needs to be in the core API (I like to keep it small, as you know ;)). The "primitive" implementation is just iterating over all keys and collecting the prefix. While this is useful, is there a real need to add this to the core API?

There is one condition under which I would add it: If it is common for backend-implementations to offer a more efficient version of this (I think S3 might do this, but they might limit it to a fixed `/` delimiter). Otherwise, it would not be possible to take advantage of these.

**Comment by crepererum** [comment]
@mbr I've added fast-paths for Azure (all possible delimiters) and FS (only works if the delimiter is the OS path separator).

For S3, this also works (see [boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.list_objects_v2) and [AWS API docs](https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html)), but I did not implement this since I think we should tackle #84 first and I'm not an S3 user.

So yes, it is rather common for upstream services to support this "directory-like" listing. I did some internal benchmarks for massive stores and you get huge advantages from the fast-path:
- Azure and S3: the reason is that you get only a limited number of keys per HTTP request (4k for Azure, 1k for S3). So if you have like 1m keys in you store but you only need to list a specific prefix, this can make a difference between 1k and 1 HTTP request, or depending on your connection the difference between 1-many minutes and a sub-second response.
- FS: the reason is just that you don't need to walk recursively through all possible directories but rather just list 1 concrete path.

**Comment by mbr** [comment]
Nice. In that case you have my blessings, I shall bikeshed no more =). Probably could not hurt to mention these things somewhere in the docs, i.e. that some backends might have better performance characteristics than others for this.

File: simplekv/__init__.py [hunk 1]
```diff
@@ -108,6 +108,33 @@ def iter_keys(self, prefix=u""):
         """
         raise NotImplementedError
 
+    def iter_prefixes_upto_delimiter(self, delimiter, prefix=u""):
```
Comment (fmarczin) [inline simplekv/__init__.py]: Would `iter_prefixes` work here as well? If so, shorter is better.

File: simplekv/decorator.py [hunk 2]
```diff
@@ -58,6 +58,20 @@ def iter_keys(self, prefix=u""):
         return (self._unmap_key(k) for k in self._dstore.iter_keys(self._map_key_prefix(prefix))
                 if self._filter(k))
 
+    def iter_prefixes_upto_delimiter(self, delimiter, prefix=u""):
```
Comment (fmarczin) [inline simplekv/decorator.py]: dito (`iter_prefixes`?)

**Comment by crepererum** [comment]
done

**Pull request merged**